File: GeneratePatterns/_code/c++/MazeGame.cpp

// Класс MazeGame отвечает за создание лабиринта.
class MazeGame {
public:
    // Метод для создания лабиринта.
    Maze* CreateMaze() {
        Maze* aMaze = new Maze;  // Создаём лабиринт
        Room* r1 = new Room(1);  // Создаём первую комнату
        Room* r2 = new Room(2);  // Создаём вторую комнату
        Door* theDoor = new Door(r1, r2);  // Создаём дверь между комнатами

        // Добавляем комнаты в лабиринт
        aMaze->AddRoom(r1);
        aMaze->AddRoom(r2);

        // Устанавливаем объекты на стороны комнаты 1
        r1->SetSide(North, new Wall);  // Северная стена
        r1->SetSide(East, theDoor);    // Восточная дверь
        r1->SetSide(South, new Wall);  // Южная стена
        r1->SetSide(West, new Wall);   // Западная стена

        // Устанавливаем объекты на стороны комнаты 2
        r2->SetSide(North, new Wall);  // Северная стена
        r2->SetSide(East, new Wall);   // Восточная стена
        r2->SetSide(South, new Wall);  // Южная стена
        r2->SetSide(West, theDoor);    // Западная дверь

        return aMaze;  // Возвращаем созданный лабиринт
    }
};

File: GeneratePatterns/_code/php/MazeGame.php
<?php

namespace App\GangOfFourDesignPatterns\GeneratePatterns\_code\php;

/**
 * Основной класс для игры, который создаёт лабиринт.
 */
class MazeGame
{
    /**
     * Метод создаёт лабиринт с двумя комнатами, соединёнными дверью.
     * Комнаты имеют стены по всем сторонам, кроме той, где расположена дверь.
     *
     * @return Maze Возвращает объект созданного лабиринта.
     */
    public function createMaze(): Maze
    {
        // Создаём объект лабиринта
        $maze = new Maze();

        // Создаём две комнаты
        $r1 = new Room(1);
        $r2 = new Room(2);

        // Создаём дверь, соединяющую две комнаты
        $door = new Door($r1, $r2);

        // Добавляем комнаты в лабиринт
        $maze->addRoom($r1);
        $maze->addRoom($r2);

        // Настраиваем стороны первой комнаты: север, восток (с дверью), юг и запад - стены
        $r1->setSide(Direction::North, new Wall);
        $r1->setSide(Direction::East, $door);
        $r1->setSide(Direction::South, new Wall);
        $r1->setSide(Direction::West, new Wall);

        // Настраиваем стороны второй комнаты: север, восток, юг и запад (с дверью)
        $r2->setSide(Direction::North, new Wall);
        $r2->setSide(Direction::East, new Wall);
        $r2->setSide(Direction::South, new Wall);
        $r2->setSide(Direction::West, $door);

        // Возвращаем готовый лабиринт
        dd($maze);
    }
}
